connect/src/com/telenor/connect/headerenrichment/HeLogic.java

```diff
@@ -0,0 +1,210 @@
+                .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+                .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
+                .build();
+        try {
+            connectivityManager.requestNetwork(
+                    networkRequest,
+                    new ConnectivityManager.NetworkCallback() {
+                        @Override
+                        public void onAvailable(Network network) {
+                            cellularNetwork = network;
+                            boolean noSignedInUser = ConnectSdk.getAccessToken() == null;
+                            if (noSignedInUser) {
+                                initializeHeaderEnrichment(useStaging, ConnectSdk.getLogSessionId());
+                            }
+                        }
+                    }
+            );
+        } catch (SecurityException e) {
+            cellularNetwork = null;
+        }
+    }
+
+    private static void initializeHeaderEnrichment(boolean useStaging, String logSessionId) {
+        if (cantDirectNetworkTraffic) {
+            return;
+        }
+
+        String url = ConnectUrlHelper.getHeApiUrl(useStaging, logSessionId);
+        GetHeaderEnrichmentGifTask getGifTask = new GetHeaderEnrichmentGifTask(url, HE_TOKEN_TIMEOUT_MILLISECONDS) {
+            @Override
+            protected void onPreExecute() {
+                HeLogic.isHeTokenRequestOngoing = true;
+                super.onPreExecute();
+            }
+
+            @Override
+            protected void onPostExecute(HeTokenResponse heToken) {
+                handleHeTokenResult(heToken);
+            }
+
+            @Override
+            protected void onCancelled(HeTokenResponse heToken) {
+                handleHeTokenResult(heToken);
+            }
+        };
+        getGifTask.execute();
+    }
+
+    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+    private static void initializeDefaultNetwork() {
+        NetworkRequest networkRequest = new NetworkRequest.Builder()
+                .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+                .build();
+        try {
+            connectivityManager.requestNetwork(
+                    networkRequest,
+                    new ConnectivityManager.NetworkCallback() {
+                        @Override
+                        public void onAvailable(Network network) {
+                            defaultNetwork = network;
+                        }
+                    }
+            );
+        } catch (SecurityException e) {
+            defaultNetwork = null;
+        }
+    }
+
+    public static void handleHeToken(
+            final Map<String, String> parameters,
+            final ShowLoadingCallback showLoadingCallback,
+            final HeTokenCallback heTokenCallback,
+            final String logSessionId,
+            final boolean useStaging) {
+        boolean finishedUnSuccessfully = !heTokenSuccess && !isHeTokenRequestOngoing;
```

Comment:
Also here is a negative parameter. Consider renaming to `finishedSuccessfully`.

Reply:
Same argument as before: never reverted.

Reply:
Ok. (As opposed to the other comment, the naming of this parameter can stay the same. The negation has its own word and camel hump, namely "Un", which makes the negation part more readable.)